File: Leetcode/Typescript/1128. Number of Equivalent Domino Pairs/solution.ts
// https://leetcode.com/problems/number-of-equivalent-domino-pairs/description

function numEquivDominoPairs(dominoes: number[][]): number {
  const hash = new Map<string, number>();
  let count = 0;

  for (const [a, b] of dominoes) {
    const key1 = `${a}${b}`;
    const key2 = `${b}${a}`;

    count += hash.get(key1) || 0;

    // Prevent duplicate
    if (a !== b) {
      count += hash.get(key2) || 0;
    }

    hash.set(key1, (hash.get(key1) || 0) + 1);
  }

  return count;
}

File: Leetcode/Python/1128. Number of Equivalent Domino Pairs/solution.py
# https://leetcode.com/problems/number-of-equivalent-domino-pairs/description

class Solution:
    def numEquivDominoPairs(self, dominoes: List[List[int]]) -> int:
        counter = defaultdict(int)
        pairs = 0

        for a, b in dominoes:
            key1 = f"{a}{b}"
            key2 = f"{b}{a}"

            pairs += counter[key1]
            
            # Prevent duplicate
            if a != b:
                pairs += counter[key2]
            
            counter[key1] += 1
        
        return pairs